gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+ *
+ * // Adding game objects to the octree
+ * octree.add(gameObject1);
+ * octree.add(gameObject2);
+ *
+ * // Querying the result
+ * ObjectSet<GameObject> result = new ObjectSet<>();
+ * octree.query(cam.frustum, result);
+ *
+ * // Rendering the result
+ * for (GameObject gameObject : result) {
+ *   modelBatch.render(gameObject);
+ * }
+ * </pre>
+ */
+public class Octree<T> {
+
+	private final int maxItemsPerNode;
+
+	private final Pool<OctreeNode> nodePool = new Pool<OctreeNode>() {
+		@Override
+		protected OctreeNode newObject() {
+			return new OctreeNode();
+		}
+	};
+
+	protected OctreeNode root;
+	private final Collider<T> collider;
+
+	private static final Vector3 tmp = new Vector3();
+
+	public Octree (Vector3 minimum, Vector3 maximum, int maxDepth, int maxItemsPerNode, Collider<T> collider) {
+		super();
+		Vector3 realMin = new Vector3(Math.min(minimum.x, maximum.x), Math.min(minimum.y, maximum.y), Math.min(minimum.z, maximum.z));
+		Vector3 realMax = new Vector3(Math.max(minimum.x, maximum.x), Math.max(minimum.y, maximum.y), Math.max(minimum.z, maximum.z));
+
+		this.root = nodePool.obtain();
+		this.root.bounds.set(realMin, realMax);
+		this.root.level = maxDepth;
+		this.collider = collider;
+		this.maxItemsPerNode = maxItemsPerNode;
+	}
+
+	public void add (T object) {
+		root.add(object);
+	}
+
+	public void remove (T object) {
+		root.remove(object);
+	}
+
+	public void update (T object) {
+		root.remove(object);
+		root.add(object);
+	}
+
+	/** Method to retrieve all the geometries.
+	 * @param resultSet
+	 * @return the result set
+	 */
+	public ObjectSet<T> getAll (ObjectSet<T> resultSet) {
+		root.getAll(resultSet);
+		return resultSet;
+	}
+
+	/** Method to query geometries inside nodes that the aabb intersects.
+	 * Can be used as broad phase.
+	 * @param aabb - The bounding box to query
+	 * @param result - Set to be populated with objects inside the BoundingBoxes
+	 */
+	public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+		root.query(aabb, result);
+		return result;
+	}
+
+	/** Method to query geometries inside nodes that the frustum intersects.
+	 * Can be used as broad phase.
+	 * @param frustum - The frustum to query
+	 * @param result set populated with objects near from the frustum
+	 */
+	public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+		root.query(frustum, result);
+		return result;
+	}
+
+	public T rayCast (Ray ray, RayCastResult<T> result) {
+		result.distance = result.maxDistanceSq;
+		root.rayCast(ray, result);
+		return result.geometry;
+	}
+
+	/** Method to get nodes as bounding boxes. Useful for debug purpose.
+	 *
+	 * @param boxes
+	 */
+	public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+		root.getBoundingBox(boxes);
+		return boxes;
+	}
+
+	protected class OctreeNode {
+
+		private int level;
+
+		// Represents the boundary of the node
+		private final BoundingBox bounds;
+		private final Array<OctreeNode> children = new Array<>(8);
```

Comment:
I suppose it could be ` new OctreeNode[8]` or null for a leaf, but this is probably fine too.

Reply:
I swear I tried that but the code doesn't compile. it says `Generic array creation`.